src/codegen/expression.rs

```diff
@@ -496,63 +496,69 @@ pub fn expression(
         ast::Expression::And(loc, left, right) => {
             and(left, cfg, contract_no, func, ns, vartab, loc, right, opt)
         }
-        ast::Expression::CheckingTrunc(loc, ty, e) => {
-            checking_trunc(loc, e, ty, cfg, contract_no, func, ns, vartab, opt)
+        ast::Expression::CheckingTrunc { loc, to, expr } => {
+            checking_trunc(loc, expr, to, cfg, contract_no, func, ns, vartab, opt)
         }
-        ast::Expression::Trunc(loc, ty, e) => Expression::Trunc(
+        ast::Expression::Trunc { loc, to, expr } => Expression::Trunc(
             *loc,
-            ty.clone(),
-            Box::new(expression(e, cfg, contract_no, func, ns, vartab, opt)),
+            to.clone(),
+            Box::new(expression(expr, cfg, contract_no, func, ns, vartab, opt)),
         ),
-        ast::Expression::ZeroExt(loc, ty, e) => Expression::ZeroExt(
+        ast::Expression::ZeroExt { loc, to, expr } => Expression::ZeroExt(
             *loc,
-            ty.clone(),
-            Box::new(expression(e, cfg, contract_no, func, ns, vartab, opt)),
+            to.clone(),
+            Box::new(expression(expr, cfg, contract_no, func, ns, vartab, opt)),
         ),
-        ast::Expression::SignExt(loc, ty, e) => Expression::SignExt(
+        ast::Expression::SignExt { loc, to, expr } => Expression::SignExt(
             *loc,
-            ty.clone(),
-            Box::new(expression(e, cfg, contract_no, func, ns, vartab, opt)),
+            to.clone(),
+            Box::new(expression(expr, cfg, contract_no, func, ns, vartab, opt)),
         ),
-        ast::Expression::Cast(loc, ty @ Type::Address(_), e) => {
-            if let Ok((_, address)) = eval_const_number(e, ns) {
-                Expression::NumberLiteral(*loc, ty.clone(), address)
+        ast::Expression::Cast { loc, to, expr } if matches!(to, Type::Address(_)) => {
```

Comment:
I think you can still write `ast::Cast{loc, to @ Type::Address(_), expr)` here.

Reply:
I thought that too, but:
```
error: expected `,`
   --> src/codegen/expression.rs:517:40
    |
517 |         ast::Expression::Cast { loc, to@ Type::Address(_), expr } => {
    |         ---------------------          ^
    |         |
    |         while parsing the fields for this pattern

```
Surprising, isn't it?